crates/networking/rpc/engine/fork_choice.rs

```diff
@@ -65,94 +70,148 @@ impl RpcHandler for ForkChoiceUpdatedV3 {
-                    parent: self.fork_choice_state.head_block_hash,
-                    timestamp: attributes.timestamp,
-                    fee_recipient: attributes.suggested_fee_recipient,
-                    random: attributes.prev_randao,
-                    withdrawals: attributes.withdrawals.clone(),
-                    beacon_root: Some(attributes.parent_beacon_block_root),
-                    version: 3,
-                };
-                let payload_id = args.id();
-                response.set_id(payload_id);
-                let payload = match create_payload(&args, &context.storage) {
-                    Ok(payload) => payload,
-                    Err(ChainError::EvmError(error)) => return Err(error.into()),
-                    // Parent block is guaranteed to be present at this point,
-                    // so the only errors that may be returned are internal storage errors
-                    Err(error) => return Err(RpcErr::Internal(error.to_string())),
-                };
-                context.storage.add_payload(payload_id, payload)?;
-            }
+            };
+            Ok((None, forkchoice_response))
         }
+    }
+}
+
+fn build_payload_v3(
+    attributes: &PayloadAttributesV3,
+    head_block: BlockHeader,
+    context: RpcApiContext,
+    fork_choice_state: &ForkChoiceState,
+) -> Result<u64, RpcErr> {
+    info!("Fork choice updated includes payload attributes. Creating a new payload.");
+    let chain_config = context.storage.get_chain_config()?;
+    if !chain_config.is_cancun_activated(attributes.timestamp) {
+        return Err(RpcErr::UnsuportedFork(
+            "forkChoiceV3 used to build pre-Cancun payload".to_string(),
+        ));
+    }
+    if attributes.timestamp <= head_block.timestamp {
+        return Err(RpcErr::InvalidPayloadAttributes(
+            "invalid timestamp".to_string(),
+        ));
+    }
+    let args = BuildPayloadArgs {
+        parent: fork_choice_state.head_block_hash,
+        timestamp: attributes.timestamp,
+        fee_recipient: attributes.suggested_fee_recipient,
+        random: attributes.prev_randao,
+        withdrawals: attributes.withdrawals.clone(),
+        beacon_root: Some(attributes.parent_beacon_block_root),
+        version: 3,
+    };
+    let payload_id = args.id();
+    let payload = match create_payload(&args, &context.storage) {
+        Ok(payload) => payload,
+        Err(ChainError::EvmError(error)) => return Err(error.into()),
+        // Parent block is guaranteed to be present at this point,
+        // so the only errors that may be returned are internal storage errors
+        Err(error) => return Err(RpcErr::Internal(error.to_string())),
+    };
+    context.storage.add_payload(payload_id, payload)?;
+
+    Ok(payload_id)
+}
+
+pub struct ForkChoiceUpdatedV2 {
+    pub fork_choice_state: ForkChoiceState,
+    pub payload_attributes: Option<PayloadAttributesV3>,
+}
+
+impl RpcHandler for ForkChoiceUpdatedV2 {
```

Comment:
A POC of a handler for `ForkChoiceUpdatedV2 `, theres a bit more of repeated code, but it is way more decoupled and has less abstractions.